################################################################################
########################     ImpulseDE2 package     ############################
################################################################################

### Version 1.0
### Author David Sebastian Fischer

################################################################################
### Libraries and source code
################################################################################

library(compiler)
library(DESeq2)
library(BiocParallel)
library(Matrix)
library(ComplexHeatmap)
library(circlize)

# Source functions in .R files from same directory as this function.
#setwd("/Users/davidsebastianfischer/gitDevelopment/ImpulseDE2/R")
#setwd("/data/yosef2/users/fischerd/code/ImpulseDE2/R")
setwd("/home/david/gitDevelopment/ImpulseDE2/R")

source("srcImpulseDE2_evalImpulse.R")
source("srcImpulseDE2_evalSigmoid.R")
source("srcImpulseDE2_compareDEMethods.R")
source("srcImpulseDE2_computeNormConst.R")
source("srcImpulseDE2_CostFunctionsFit.R")
source("srcImpulseDE2_fitImpulse.R")
source("srcImpulseDE2_fitSigmoid.R")
source("srcImpulseDE2_plotDEGenes.R")
source("srcImpulseDE2_processData.R")
source("srcImpulseDE2_runDEAnalysis.R")
source("srcImpulseDE2_runDESeq2.R")
source("srcImpulseDE2_simulateDataSet.R")

# Compile function
evalImpulse_comp <- cmpfun(evalImpulse)
evalSigmoid_comp <- cmpfun(evalSigmoid)
evalLogLikMu_comp <- cmpfun(evalLogLikMu)
evalLogLikImpulse_comp <- cmpfun(evalLogLikImpulse)
evalLogLikSigmoid_comp <- cmpfun(evalLogLikSigmoid)

################################################################################
### Main function
################################################################################

#' Differential expression analysis using impulse models
#'
#' Fits an impulse model to time course data and uses this model as a basis
#' to detect differentially expressed genes. Differential expression is
#' either differential expression of a gene over time within one condition
#' or differential expression of a gene over time between two conditions (
#' case and control). With a single condition, the alternative model is the
#' impulse fit to the time course data and the null model is the mean fit.
#' The mean fit models no differential behaviour over time. With two 
#' conditions, the alternative model is separate impulse fits to case and
#' control data and the null model is an impulse fit to the combined data.
#' Here, the impulse fit to the combined data models no differential expression
#' between the conditions. ImpulseDE2 can operate on bulk omics count data
#' or on scRNA-seq counts. In the latter case additional hyperparameters
#' have to be supplied, which can be done by using the wrapper supplied
#' in LineagePulse.
#' To run ImpulseDE2 on bulk omics data, use the minimal
#' parameter set matCountData, dfAnnotation, strCaseName, strControlName,
#' strMode. Use nProc to set the number of processes assigned and 
#' Q_value to set the cut off for your DE gene list. Set the plotting behaviour
#' with boolPlotting, boolSimplePlot and boolLogPlot or leave as defaults.
#' Supply dispersions and size factors externally if you wish through
#' vecDispersionsExternal and vecSizeFactorsExternal, these are otherwise
#' generated by ImpulseDE2. Dispersions are generated with DESeq2, if you supply
#' dispersions externally and do not wish to have reference p-values from
#' DESeq2, you can turn off the running of DESeq2 with boolRunDESeq2. If you 
#' analyse scRNA-seq data, you need the additional parameters strSCMode,
#' scaWindowRadius and lsPseudoDE. You can test-run ImpulseDE2 on
#' a subset of your data to make sure that your input works or to get an
#' idea of what the output looks like with scaSmallRun. By default,
#' temporary data are saved to the working directory as .RData, which can be
#' turned off with boolSaveTemp.
#' 
#' @details \code{ImpulseDE2} is based on the impulse model proposed by
#' Chechik and Koller (Chechik and Koller, 2009). The impulse model
#' models the response of gene activity read outs (such as RNAseq counts)
#' to environmental or developmental stimuli as the product
#' of two simgoids. This model can capture simple time course
#' patterns, such as plateaus, increase and decrease. ImpulseDE2 uses
#' the impulse model to identify differential activity over time on any
#' type of count data which follows the negative binomial distribution
#' (as frequently encountered in sequencing data). ImpulseDE2 performs
#' fitting of the impulse model and a mean model to data and evaluates
#' the fit. The computational complexity of ImpulseDE2 is linear in the
#' number of genes and linear in the number of samples.
#' \enumerate{
#'  \item Impulse fitting: The impulse model is fitted based on the assumption
#'    that the input count data follow a negative binomial distribution with
#'    dispersion as identified by DESeq2. The impulse model does not have
#'    a closed form maximum likelihood parameter estimate and must therefore
#'    be inferred from numerical optimisation.
#'  \enumerate{
#'    \item Initialisation: Initialisation is performed twice for each gene,
#'      based on a peak and a valley model. The parameters representing these 
#'      models reflect the form that these two models would have given the count
#'      data of each gene and are specific to each gene.
#'    \item Optimisation: The cost function for the fit is the log likelihood
#'      of the data which is evaluated based on negative binomial likelihoods
#'      at each observed time point, with the value of the impulse model as the
#'      mean and the gene dispersion inferred using DESeq2 as the 
#'      dispersion. Numerical optimisation is performed using the
#'      BFGS algorithm. In analogy to generalised linear models for count data,
#'      the fitting of the parameters representing limit behaviours of the two
#'      sigmoids (which are counts) are fitted in log space so that they cannot
#'      adopt negative values.
#'    \item Fit selection: The fit with the higher log likelihood of the two
#'      initialisation is selected and kept as a maximum likelihood estimate.
#'  }
#'  \item Mean fitting: The mean model is a single negative binomial and
#'    serves as the null model in the case of differential expression over
#'    time within a single condition. The maximum likelihood estimate of
#'    the mean of a negative binomial distribution is computed.
#'  \item Fit evaluation: The model comparison statistic is the deviance
#'    2 * (loglikelihood(H1) - loglikelihood(H0)). The deviance is chi-squared
#'    distributed with the difference in degrees of freedom of both models
#'    as degrees of freedom. Therefore p-values for differential expression 
#'    are computed based on the chi-squared distribution.
#'    The p-values are the FDR corrected (Benjamini and Hochberg, 1995).
#' }
#' 
#' @aliases ImpulseDE2
#' 
#' @param matCountData: (matrix genes x samples) [Default NULL] 
#'    Read count data, unobserved entries are NA.
#'    Read count data.
#' @param dfAnnotationProc: (Table samples x covariates) 
#'    {Sample, Condition, Time (numeric), TimeCateg (str)
#'    (and confounding variables if given).}
#'    Annotation table with covariates for each sample.
#' @param boolCaseCtrl: (bool) 
#' 		Whether to perform case-control analysis. Does case-only
#' 		analysis if FALSE.
#' @param vecConfounders: (vector of strings number of confounding variables)
#' 		Factors to correct for during batch correction. Have to 
#' 		supply dispersion factors if more than one is supplied.
#' 		Names refer to columns in dfAnnotation.
#' @param scaNProc: (scalar) [Default 1] Number of processes for 
#'    parallelisation.
#' @param scaQThres: (scalar) [Default 0.01] 
#'    FDR-corrected p-value cutoff for significance.
#' @param vecDispersionsExternal: (vector length number of
#'    genes in matCountData) [Default NULL]
#'    Externally generated list of gene-wise dispersion factors
#'    which overides DESeq2 generated dispersion factors.
#' @param vecSizeFactorsExternal: (vector length number of
#'    cells in matCountData) [Default NULL]
#'    Externally generated list of size factors which override
#'    size factor computation in ImpulseDE2.
#' @param boolPlotting: (bool) [TRUE] 
#'    Whether to plot significant DE genes into output pdf.
#'    Consider setting FALSE for large data sets with many hits.
#' @param boolSimplePlot: (bool) [Default FALSE]
#'    Whether to reduce plot to data points and impulse trace.
#' @param dirTemp: (dir) Directory to which temporary results are saved.
#' 
#' @return (list length 4)
#' \itemize{
#'    \item vecDEGenes: (list number of genes) Genes IDs identified
#'    as differentially expressed by ImpulseDE2 at threshold \code{scaQThres}.
#'    \item dfImpulseResults: (data frame) ImpulseDE2 results.
#'    \item lsModelFits: (list length number of conditions fit (1 or 3))
#'    {"case"} or {"case", "control", "combined"}
#'    One model fitting object for each condition:
#'    In case-only DE analysis, only the condition {"case"} is fit.
#'    In case-control DE analysis, the conditions 
#'    {"case", "control","combined} are fit.
#'    Each condition entry is a list of model fits for each gene.
#'    Each gene entry is a list of model fits to the individual models:
#'    Impulse model and constant model (if boolFitConst is TRUE).
#'    At this level, the sigmoid model fit can be added later.
#'    Each model fit per gene is a list of fitting parameters and results.
#'    \itemize{
#'      \item Condition ID: (list length number of genes)
#'      List of fits for each gene to the samples of this condition.
#'      One entry of this format for all conditions fit.
#'      \itemize{
#'        \item Gene ID: (list length 2)
#'        Impulse and constant model fit to gene observations.
#'        One entry of this format for all gene IDs.
#'        \itemize{
#'          \item lsImpulseFit: (list) List of impulse fit parameters and results.
#'          \itemize{
#'            \item vecImpulseParam: (numeric vector length 6)
#'            {beta, h0, h1, h2, t1, t2}
#'            Maximum likelihood estimators of impulse model parameters.
#'            \item vecImpulseValue: (numeric vector length number of time points)
#'            Values of impulse model fit at time points used for fit.
#'            \item lsvecBatchFactors: (list length number of confounders)
#'            List of vectors of scalar batch correction factors for each sample.
#'            These are also maximum likelihood estimators.
#'            NULL if no confounders given.
#'            \item scaDispParam: (scalar) Dispersion parameter estimate
#'            used in fitting (hyper-parameter).
#'            \item scaLL: (scalar) Loglikelihood of data under maximum likelihood
#'            estimator model.
#'            \item scaConvergence: (scalar) 
#'            Convergence status of optim on impulse model.
#'          }
#'          \item lsConstFit: (list) List of constant fit parameters and results.
#'          \itemize{
#'            \item scaMu: (scalar) Maximum likelihood estimator of
#'            negative binomial mean parameter.
#'            \item lsvecBatchFactors: (list length number of confounders)
#'            List of vectors of scalar batch correction factors for each sample.
#'            These are also maximum likelihood estimators.
#'            NULL if no confounders given.
#'            \item scaDispParam: (scalar) Dispersion parameter estimate
#'            used in fitting (hyper-parameter).
#'            \item scaLL: (scalar) Loglikelihood of data under maximum likelihood
#'            estimator model.
#'            \item scaConvergence: (scalar) 
#'            Convergence status of optim on constant model.
#'          }
#'          \item ls SigmoidFit: (list) List of sigmoidal fit parameters and results.
#'          NULL if boolIdentifyTransients is FALSE.
#'          \itemize{
#'            \item vecSigmoidParam: (numeric vector length 4)
#'            {beta, h0, h1, t}
#'            Maximum likelihood estimators of sigmoidal model parameters.
#'            \item vecSigmoidValue: (numeric vector length number of time points)
#'            Values of sigmoid model fit at time points used for fit.
#'            \item lsvecBatchFactors: (list length number of confounders)
#'            List of vectors of scalar batch correction factors for each sample.
#'            These are also maximum likelihood estimators.
#'            NULL if no confounders given.
#'            \item scaDispParam: (scalar) Dispersion parameter estimate
#'            used in fitting (hyper-parameter).
#'            \item scaLL: (scalar) Loglikelihood of data under maximum likelihood
#'            estimator model.
#'            \item scaConvergence: (scalar) 
#'            Convergence status of optim on sigmoidal model.
#'          }
#'        }
#'      }
#'    }
#' }
#' Additionally, \code{ImpulseDE2} saves the following objects and tables into
#' the working directory:
#' \itemize{
#'    \item \code{ImpulseDE2_matCountDataProc.RData} (2D array genes x replicates) 
#'        Count data: Reduced version of \code{matCountData}. For internal use.
#'    \item \code{ImpulseDE2_dfAnnotationProc.RData} (data frame) Annotation table.
#'    \item \code{ImpulseDE2_vecSizeFactors.RData} (numeric vector number of samples) 
#'        Model scaling factors for each observation which take
#'        sequencing depth into account (size factors). One size
#'        factor per sample.
#'    \item \code{ImpulseDE2_vecDispersions.RData} (vector number of genes) Inverse 
#'        of gene-wise negative binomial dispersion coefficients computed by DESeq2.
#'    \item \code{ImpulseDE2_lsModelFits.RData} (list length number of conditions fit (1 or 3))
#'    {"case"} or {"case", "control", "combined"}
#'    One model fitting object for each condition:
#'    In case-only DE analysis, only the condition {"case"} is fit.
#'    In case-control DE analysis, the conditions 
#'    {"case", "control","combined} are fit.
#'    Each condition entry is a list of model fits for each gene.
#'    Each gene entry is a list of model fits to the individual models:
#'    Impulse model and constant model (if boolFitConst is TRUE).
#'    At this level, the sigmoid model fit can be added later.
#'    Each model fit per gene is a list of fitting parameters and results.
#'    Read the return-value section of runImpulseDE2 for details.
#'    \item \code{ImpulseDE2_dfImpulseDE2Results.RData} (data frame) ImpulseDE2 results.
#'    \item \code{ImpulseDE2_vecDEGenes.RData} vecDEGenes: (list number of genes) Genes IDs identified
#'    as differentially expressed by ImpulseDE2 at threshold \code{scaQThres}.
#' 
#' @seealso Calls the following functions:
#' \code{\link{processData}}, 
#' \code{\link{runDESeq2}},
#' \code{\link{fitImpulse}},
#' \code{\link{computePval}}, 
#' \code{\link{plotDEGenes}}.
#' 
#' @author David Sebastian Fischer
#' 
#' @export

runImpulseDE2 <- function(matCountData=NULL, 
  dfAnnotation=NULL,
  boolCaseCtrl=NULL,
  vecConfounders=NULL,
  scaNProc=1, 
  scaQThres=0.01,
  vecDispersionsExternal=NULL,
  vecSizeFactorsExternal=NULL,
  boolIdentifyTransients=FALSE,
  boolPlotting=FALSE,
  boolSimplePlot=FALSE,
  dirTemp=NULL ){
  
  print("ImpulseDE2 v1.0 for count data")
  
  tm_runImpulseDE2 <- system.time({    
    # 1. Process input data 
    print("# Process input")
    lsProcessedData <- processData(
      dfAnnotation=dfAnnotation,
      matCountData=matCountData,
      boolCaseCtrl=boolCaseCtrl, 
      vecConfounders=vecConfounders,
      vecDispersionsExternal=vecDispersionsExternal,
      vecSizeFactorsExternal=vecSizeFactorsExternal)
    
    matCountDataProc <- lsProcessedData$matCountDataProc
    dfAnnotationProc <- lsProcessedData$dfAnnotationProc
    vecSizeFactorsExternalProc <- lsProcessedData$vecSizeFactorsExternalProc
    
    # Set parallelisation
    if(scaNProc > 1){ register(MulticoreParam(workers=scaNProc)) 
    } else { register(SerialParam()) }
    
    if(!is.null(dirTemp)){
      save(matCountDataProc,file=file.path(dirTemp,"ImpulseDE2_matCountDataProc.RData"))
      save(dfAnnotationProc,file=file.path(dirTemp,"ImpulseDE2_dfAnnotationProc.RData"))
    }
    
    # 2. Run DESeq2
    # Use dispersion factors from DESeq2 if
    if(is.null(vecDispersionsExternal)){
      print("# Run DESeq2: Using dispersion factors computed by DESeq2.")
      tm_runDESeq2 <- system.time({
        vecDispersions <- runDESeq2(
          dfAnnotationProc=dfAnnotationProc,
          matCountDataProc=matCountDataProc,
          boolCaseCtrl=boolCaseCtrl,
          vecConfounders=vecConfounders)
      })
      print(paste0("Consumed time: ",round(tm_runDESeq2["elapsed"]/60,2), " min."))
    } else {
      # Use externally provided dispersions and reorder
      print("# Using externally supplied dispersion factors.")
      vecDispersions <- vecDispersionsExternal
    }
    if(!is.null(dirTemp)){
      save(vecDispersions,file=file.path(dirTemp,"ImpulseDE2_vecDispersions.RData"))
    }
    
    # 3. Compute size factors
    print("# Compute size factors")
    vecSizeFactors <- computeNormConst(
    	matCountDataProc=matCountDataProc,
      vecSizeFactorsExternal=vecSizeFactorsExternalProc )
    if(!is.null(dirTemp)){
      save(vecSizeFactors,file=file.path(dirTemp,"ImpulseDE2_vecSizeFactors.RData"))
    }
    
    ###  4. Fit null and alternative model to each gene
    print("# Fitting null and alternative model to the genes")
    tm_fitImpulse <- system.time({
      lsModelFits <- fitModels(
        matCountDataProc=matCountDataProc, 
        vecDispersions=vecDispersions,
        vecSizeFactors=vecSizeFactors,
        dfAnnotationProc=dfAnnotationProc,
        vecConfounders=vecConfounders,
        boolCaseCtrl=boolCaseCtrl)
    })
    if(!is.null(dirTemp)){
      save(lsModelFits,file=file.path(dirTemp,"ImpulseDE2_lsModelFits.RData"))
    }
    print(paste0("Consumed time: ",round(tm_fitImpulse["elapsed"]/60,2)," min."))
    
    ###  5. Fit sigmoid model to case condition if desired
    if(boolIdentifyTransients){
      print("# Fitting sigmoid model to case condition")
      tm_fitSigmoid <- system.time({
        lsModelFits <- fitSigmoidModels(
          matCountDataProc=matCountDataProc,
          lsModelFits=lsModelFits,
          vecDispersions=vecDispersions,
          vecSizeFactors=vecSizeFactors,
          dfAnnotationProc=dfAnnotationProc,
          vecConfounders=vecConfounders,
          strCondition="case")
      })
      if(!is.null(dirTemp)){
        save(lsModelFits,file=file.path(dirTemp,"ImpulseDE2_lsModelFits.RData"))
      }
      print(paste0("Consumed time: ",round(tm_fitSigmoid["elapsed"]/60,2)," min."))
    }
    
    ### 6. Differentially expression analysis based on model fits
    print("# Differentially expression analysis based on model fits")
    dfImpulseDE2Results <- runDEAnalysis(
      matCountDataProc=matCountDataProc,
      dfAnnotationProc=dfAnnotationProc,
      lsModelFits=lsModelFits,
      boolCaseCtrl=boolCaseCtrl,
      vecConfounders=vecConfounders,
      boolIdentifyTransients=boolIdentifyTransients)
    
    if(!is.null(scaQThres)){
      vecDEGenes <- as.vector( 
        dfImpulseDE2Results[as.numeric(dfImpulseDE2Results$padj) <= scaQThres,"Gene"] )
      print(paste0("Found ", length(vecDEGenes)," DE genes",
        " at a FDR corrected p-value cut off of ", scaQThres, "."))
    } else {
      vecDEGenes <- NULL
    }
    if(!is.null(dirTemp)){
      save(dfImpulseDE2Results,file=file.path(dirTemp,"ImpulseDE2_dfImpulseDE2Results.RData"))
      if(!is.null(scaQThres)){
        save(vecDEGenes,file=file.path(dirTemp,"ImpulseDE2_vecDEGenes.RData"))
      }
    }
    
    ### 7. Plot differentially expressed genes
    if(boolPlotting){
      print("# Plot differentially expressed genes")
      if(length(vecDEGenes) > 500){vecDEGenesPlot <- vecDEGenes[1:500]
      } else {vecDEGenesPlot <- vecDEGenes}
      tm_plotDEGenes <- system.time({
        plotDEGenes(
          vecGeneIDs=vecDEGenesPlot,
          matCountDataProc=matCountDataProc,
          matBatchFactors=matBatchFactors,
          matSizeFactors=matSizeFactors,
          dfAnnotationProc=dfAnnotationProc, 
          lsModelFits=lsModelFits,
          dfImpulseDE2Results=dfImpulseDE2Results,
          vecRefPval=vecRefPval,
          strMode=strMode,
          strCaseName=strCaseName, 
          strControlName=strControlName, 
          strFileNameSuffix="DEgenes", 
          strPlotTitleSuffix="", 
          strPlotSubtitle="",
          strNameMethod2=strRefMethod,
          boolSimplePlot=boolSimplePlot,
          boolLogPlot=boolLogPlot)
      })
      print(paste0("Consumed time: ",round(tm_plotDEGenes["elapsed"]/60,2)," min."))
    }
  })
  print("Finished running ImpulseDE2.")
  print(paste0("TOTAL consumed time: ",round(tm_runImpulseDE2["elapsed"]/60,2)," min."))
  
  return(list(
    dfImpulseDE2Results=dfImpulseDE2Results,
    vecDEGenes=vecDEGenes,
    lsModelFits=lsModelFits
  ))
}
